ArchiSteamFarm/Core/AprilFools.cs

```diff
@@ -0,0 +1,83 @@
+﻿//     _                _      _  ____   _                           _____
+//    / \    _ __  ___ | |__  (_)/ ___| | |_  ___   __ _  _ __ ___  |  ___|__ _  _ __  _ __ ___
+//   / _ \  | '__|/ __|| '_ \ | |\___ \ | __|/ _ \ / _` || '_ ` _ \ | |_  / _` || '__|| '_ ` _ \
+//  / ___ \ | |  | (__ | | | || | ___) || |_|  __/| (_| || | | | | ||  _|| (_| || |   | | | | | |
+// /_/   \_\|_|   \___||_| |_||_||____/  \__|\___| \__,_||_| |_| |_||_|   \__,_||_|   |_| |_| |_|
+// |
+// Copyright 2015-2021 Łukasz "JustArchi" Domeradzki
+// Contact: JustArchi@JustArchi.net
+// |
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// |
+// http://www.apache.org/licenses/LICENSE-2.0
+// |
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Globalization;
+using System.Threading;
+
+namespace ArchiSteamFarm.Core {
+	internal static class AprilFools {
+		private static readonly object LockObject = new();
+
+		// We don't care about CurrentCulture global config property, because April Fools are never initialized in this case
+		private static readonly CultureInfo OriginalCulture = CultureInfo.CurrentCulture;
+
+		private static readonly Timer Timer = new(Init);
+
+		internal static void Init(object? state = null) {
+			DateTime now = DateTime.Now;
+
+			if ((now.Month == 4) && (now.Day == 1)) {
+				try {
+					CultureInfo.DefaultThreadCurrentCulture = CultureInfo.DefaultThreadCurrentUICulture = CultureInfo.CreateSpecificCulture("qps-Ploc");
+				} catch (Exception e) {
+					ASF.ArchiLogger.LogGenericDebuggingException(e);
+
+					return;
+				}
+
+				TimeSpan aprilFoolsEnd = TimeSpan.FromDays(1) - now.TimeOfDay;
+
+				lock (LockObject) {
+					Timer.Change(aprilFoolsEnd, Timeout.InfiniteTimeSpan);
+				}
+
+				return;
+			}
+
+			try {
+				CultureInfo.DefaultThreadCurrentCulture = CultureInfo.DefaultThreadCurrentUICulture = OriginalCulture;
+			} catch (Exception e) {
+				ASF.ArchiLogger.LogGenericDebuggingException(e);
+
+				return;
+			}
+
+			// Since we already verified that it's not April Fools right now, either we're in months 1-3 before 1st April this year, or 4-12 already after the 1st April
+			DateTime nextAprilFools = new(now.Month >= 4 ? now.Year + 1 : now.Year, 4, 1, 0, 0, 0, DateTimeKind.Local);
+
+			TimeSpan aprilFoolsStart = nextAprilFools - now;
+
+			// Timer can accept only dueTimes up to 2^32 - 2
+			uint dueTime = (uint) Math.Min(uint.MaxValue - 1, (ulong) aprilFoolsStart.TotalMilliseconds);
```

Comment:
Have you tested the switch at runtime? I can see some edge case, where the timer is executed at the exact millisecond the date rolls over but `DateTime.Now` is slightly inaccurate. As it currently is, we might want to add 100 milliseconds (or a bit more) of room for error (see line 160 in referenced source code) to `aprilFoolsStart.TotalMilliseconds`.

reference:
[dotnet runtime source](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs#L143) (issue is linked in comment)

Reply:
This is not an issue, there are 3 scenarios:
- Timer executes too early, nothing will happen, another timer will be scheduled in several miliseconds from now (valid)
- Timer executes just in time, logic will happen as expected
- Timer executes too late, logic will happen as expected

Notice how there is a reason why I calculate `DateTime.Now` exactly **once**, and then base the scenario off that, so it doesn't switch between one calculation and the other. Either we hit the case we expected (so the logic goes off), or we don't, and if we don't we'll reschedule the timer according to the logic which reschedules the timer in all other scenarios.

Example:

```
31 March, 23:59:00, timer executed in 60s
31 March 23:59:59.700, timer executed in 300ms
31 March 23:59:59.995, timer executed in 5ms
1 April 00:00:00.010, logic called, timer executed in 23h 59m 59s 990ms
1 April 23:59:59.800, timer executed in 200 ms
2 April 00:00:00.100, logic called, timer executed in ~49 days or whatever max is
```

This is a VERY unlikely scenario because keep in mind that overhead that comes from rescheduling the timer, executing the function, calculating current time and everything else the function is doing almost guarantees that at most we'll have one reschedule per logic call expectation, as the closer we are to the result, the more precise the timer is, and there are already a few miliseconds spent on evaluating whether the time is right.

Let me know if I missed something, but I don't see it.

Reply:
But you're right, adding additional second or so should not hurt this logic more than needed, and if it decreases chance for additional reschedule logic then we save CPU cycles, so adding a few hundred ms sounds like a good plan.

Reply:
Since the lowest tickrate for CPUs is `100 Hz`, this means 1 tick per 33.3ms, 100 ms, so 3 ticks should do the trick.

Reply:
You are right, it will reschedule the timer a few milliseconds from "now" in the edge case I thought about and I fully understood that when writing the comment. It was just an idea that I wanted to throw into the room to avoid that useless reschedule at all by adding a tiny offset.